### PR TITLE
Fixed an issue wher case timeout returned an incorrect exception

### PR DESCRIPTION
**Pull request by hzsEmperor**
Fixes 
The current judgment condition causes the exception result to InterruptedException instead of the expected ThreadTimeoutException after the method call timeout is interrupted.

So I call Thread.interrupted() to ensure that the correct exception type is returned, while restoring the interrupt status value of Thread so that its subsequent operations are unaffected.

If you want to reproduce this problem, please make sure to create a new TestNG object through Java and set ParallelMode to any non-Tests

### CODE REVIEW

**Comment by juherr** [comment]
Hi,

Thanks for your contribution.

Could you add a test case for non-regression? 
Also, it will allow us to understand clearly what you are trying to do and determine if your fix is the best way to do it.

**Review by krmahadevan — CHANGES_REQUESTED** [review]
Also there are test failures to be fixed:

* `test.expectedexceptions.issue2074.IssueTest.testIfInterruptsAreCleared`
* `test.expectedexceptions.issue2235.IssueTest.testExpectedExceptions`

**Comment by hzsEmperor** [comment]
Hi,
Just got a new bug here!
The method "invokeWithTimeoutWithNoExecutor" creates a new thread, sleep the specified time and interrupt the current thread to guarantee timeout.
However, if the execution time of this method is less than the time set for timeout, the next method will be executed after completion, and then the new thread will interrupt the current thread directly after the end of sleep, resulting in an interrupt exception.

This bug really bothered me and kept me from having a good rest on weekends！！！

**Comment by hzsEmperor** [comment]
> Also there are test failures to be fixed:
> 
> * `test.expectedexceptions.issue2074.IssueTest.testIfInterruptsAreCleared`
> * `test.expectedexceptions.issue2235.IssueTest.testExpectedExceptions`

The exceptions for these two cases are in my changes, but I avoided this in order to get them through

**Comment by hzsEmperor** [comment]
> Hi,
> 
> Thanks for your contribution.
> 
> Could you add a test case for non-regression?
> Also, it will allow us to understand clearly what you are trying to do and determine if your fix is the best way to do it.

Run this, you will see 
java.lang.InterruptedException: sleep interrupted
instead of org.testng.internal.thread.ThreadTimeoutException
`
public class DemoTest extends TestListenerAdapter {

    public static void main(String[] args) {
        TestNG testNG = new TestNG();
        XmlSuite xmlSuite = new XmlSuite();
        XmlTest xmlTest = new XmlTest(xmlSuite);
        xmlTest.setClasses(Collections.singletonList(new XmlClass(DemoTest.class)));
        xmlSuite.setParallel(XmlSuite.ParallelMode.CLASSES);
        xmlSuite.setTests(Collections.singletonList(xmlTest));
        testNG.setListenerClasses(Collections.singletonList(DemoTest.class));
        testNG.setXmlSuites(Collections.singletonList(xmlSuite));
        testNG.run();
    }

    @Test(description = "Demo测试用例", timeOut = 3)
    public void demo3Test() throws InterruptedException {
        Thread.sleep(7200);
    }

    @Override
    public void onTestFailure(ITestResult iTestResult) {
        iTestResult.getThrowable().printStackTrace();
    }
}
`

**Comment by hzsEmperor** [comment]
> Hi,
> Just got a new bug here!
> The method "invokeWithTimeoutWithNoExecutor" creates a new thread, sleep the specified time and interrupt the current thread to guarantee timeout.
> However, if the execution time of this method is less than the time set for timeout, the next method will be executed after completion, and then the new thread will interrupt the current thread directly after the end of sleep, resulting in an interrupt exception.
> 
> This bug really bothered me and kept me from having a good rest on weekends！！！


Run this you will see 
"Demo222222
java.lang.InterruptedException: sleep interrupted"
```
public class DemoTest extends TestListenerAdapter {

    public static void main(String[] args) {
        TestNG testNG = new TestNG();
        XmlSuite xmlSuite = new XmlSuite();
        XmlTest xmlTest = new XmlTest(xmlSuite);
        xmlTest.setClasses(Collections.singletonList(new XmlClass(DemoTest.class)));
        xmlSuite.setParallel(XmlSuite.ParallelMode.CLASSES);
        xmlSuite.setTests(Collections.singletonList(xmlTest));
        testNG.setListenerClasses(Collections.singletonList(DemoTest.class));
        testNG.setXmlSuites(Collections.singletonList(xmlSuite));
        testNG.run();
    }

    @Test(description = "Demo", timeOut = 1000)
    public void demo3Test() throws InterruptedException {
        Thread.sleep(500);
    }

    @Test(description = "Demo222222")
    public void demo4Test() throws InterruptedException {
        Thread.sleep(7200);
    }

    @Override
    public void onTestFailure(ITestResult iTestResult) {
        System.out.println(iTestResult.getMethod().getDescription());
        iTestResult.getThrowable().printStackTrace();
    }
}
```

**Comment by juherr** [comment]
@hzsEmperor You have samples, just convert them into runnable TestNG test cases :) 

> The exceptions for these two cases are in my changes, but I avoided this in order to get them through

How do you suggest to fix them?

**Comment by hzsEmperor** [comment]
> @hzsEmperor You have samples, just convert them into runnable TestNG test cases :)
But i don't want to! I have my own job!!!

> > The exceptions for these two cases are in my changes, but I avoided this in order to get them through
Forget this!!!

> How do you suggest to fix them?
Which in my commit!!!!

Don't make it so hard, I just want to submit my code and fix your bugs.

I have given you the example, if you want a new test case, you can implement it yourself (by the way, it may cause an out of Memory exception, which caused my MBP to shut down automatically twice!).

Do me a favor, look at this method, then look at my COMMIT, few lines of code, very easy to understand!!

**Comment by juherr** [comment]
You are funny :)

**Comment by juherr** [comment]
> Is this a necessary step?

Like you, we have our own job and we can't spend too much time with incomplete propositions.
So yes, it is mandatory (and mostly to assure non-regression in the future).

About your fix itself, it looks nice but it breaks existing tests and we can't merge without a green light on tests.
The question is why are they failing? How are they supposed to be fixed (and not removed!)?
Sadly, we have the same amount of free time and it won't be possible to dig too much on our side for the moment.
 
I hope you'll understand.

**Review by krmahadevan — CHANGES_REQUESTED** [review]
@hzsEmperor - I dont have anything additional to add apart from what @juherr already called out. Please help with 

* Adding relevant tests that vet out if the proposed changes work fine.
* Fix any failing tests so that we know that we havent broken any working functionality.

Also FYI., I believe its a basic thing to expect a PR to contain relevant tests baked into it (only exception would be documentation changes). And no we won't be able to devote the time to add a test to your PR. Whenever you do find the time, please add the tests and update your PR and we will be more than glad to help with the merge. 

Fixing an issue doesn't necessarily mean just code changes. It also includes unit tests that re-iterate that the changes do work fine the way they are intended to.

Thanks in advance for your understanding and co-operation in this regard.

**Comment by juherr** [comment]
Replaced by #2444

**Pull request closed without merging**
